library/psa_crypto.c

```diff
@@ -6288,30 +6310,126 @@ psa_status_t psa_raw_key_agreement( psa_algorithm_t alg,
+ */
+static psa_status_t mbedtls_psa_random_seed( mbedtls_psa_random_context_t *rng )
+{
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+    /* Do nothing: the external RNG seeds itself. */
+    (void) rng;
+    return( PSA_SUCCESS );
+#else /* MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+    const unsigned char drbg_seed[] = "PSA";
+    int ret = mbedtls_psa_drbg_seed( &rng->entropy,
+                                     drbg_seed, sizeof( drbg_seed ) - 1 );
+    return mbedtls_to_psa_error( ret );
+#endif /* MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+}
+
 psa_status_t psa_generate_random( uint8_t *output,
                                   size_t output_size )
 {
-    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     GUARD_MODULE_INITIALIZED;
 
-    while( output_size > MBEDTLS_CTR_DRBG_MAX_REQUEST )
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+
+    size_t output_length = 0;
+    psa_status_t status = mbedtls_psa_external_get_random( &global_data.rng,
+                                                           output, output_size,
+                                                           &output_length );
+    if( status != PSA_SUCCESS )
+        return( status );
+    /* Breaking up a request into smaller chunks is currently not supported
+     * for the extrernal RNG interface. */
+    if( output_length != output_size )
+        return( PSA_ERROR_INSUFFICIENT_ENTROPY );
+    return( PSA_SUCCESS );
+
+#else /* MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+
+    while( output_size > MBEDTLS_PSA_RANDOM_MAX_REQUEST )
     {
-        ret = mbedtls_ctr_drbg_random( &global_data.ctr_drbg,
-                                       output,
-                                       MBEDTLS_CTR_DRBG_MAX_REQUEST );
+        ret = mbedtls_psa_get_random(
+            MBEDTLS_PSA_RANDOM_STATE,
+            output, MBEDTLS_PSA_RANDOM_MAX_REQUEST );
         if( ret != 0 )
             return( mbedtls_to_psa_error( ret ) );
-        output += MBEDTLS_CTR_DRBG_MAX_REQUEST;
-        output_size -= MBEDTLS_CTR_DRBG_MAX_REQUEST;
+        output += MBEDTLS_PSA_RANDOM_MAX_REQUEST;
+        output_size -= MBEDTLS_PSA_RANDOM_MAX_REQUEST;
     }
 
-    ret = mbedtls_ctr_drbg_random( &global_data.ctr_drbg, output, output_size );
+    ret = mbedtls_psa_get_random( MBEDTLS_PSA_RANDOM_STATE,
+                                  output, output_size );
     return( mbedtls_to_psa_error( ret ) );
+#endif /* MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG */
+}
+
+/* Wrapper function allowing the classic API to use the PSA RNG.
+ * In the non-external case, mbedtls_psa_get_random is defined
+ * as a constant function pointer in psa_crypto_random_impl.h.
+ */
+#if defined (MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+int mbedtls_psa_get_random( void *p_rng,
+                            unsigned char *output,
+                            size_t output_size )
+{
+    (void) p_rng;
+    psa_status_t status = psa_generate_random( output, output_size );
```

Comment:
I don't want to specialize `mbedtls_psa_get_random` because that won't work when we add full support for the RNG accelerator interface. It would sacrifice the cleanliness of the high-level design for the sake of having a slightly simpler code path, and the change would need to be undone in a few months.

Reply:
I don't understand "I don't want to specialize `mbedtls_psa_get_random()`". What do you mean by specialization? What do you mean by "It would sacrifice the cleanliness of the high-level design" ?

Reply:
High-level design: getting random data can be done either through a PSA driver or through classic mbedtls code. As a consequence, the PSA API function `psa_get_random` has two cases in its implementation. Furthermore, the internal function `mbedtls_psa_get_random` either calls the PSA API function if PSA is the “best” way to do what it does, or calls a classic mbedtls function if PSA isn't “best”. PSA is “best” iff there's a PSA driver.

In general, a PSA random driver can take many forms, depending on whether the driver is the exclusive source of entropy, one source among several, or is only a PRNG driver. At this point, we're only implementing the case of a driver that is the exclusive source of entropy. You want to make `mbedtls_psa_get_random` less generic in the case of a PSA driver, by making it specialized to the case where the driver is the exclusive source of entropy. I don't want to do this.